#include "mainwindow.h"
#include "./ui_mainwindow.h"

#include <QString>
#include <QFileDialog>
#include <QDir>
#include <QInputDialog>
#include <QMessageBox>
#include <QCryptographicHash>
#include <aclapi.h>
#include <winerror.h>

QString filePath = NULL;

MainWindow::MainWindow(QWidget *parent)
    : QMainWindow(parent)
    , ui(new Ui::MainWindow)
{
    filePath = QCoreApplication::applicationDirPath();
    ui->setupUi(this);
    ui->lineEdit->setText(filePath);

    read_from_template();
}

MainWindow::~MainWindow()
{
    delete ui;
}


void MainWindow::on_pushButton_clicked()
{
    QString tempDirName = QFileDialog::getExistingDirectory(this, "Choose Directory", filePath,QFileDialog::ShowDirsOnly);
    if(tempDirName.isEmpty()){
        //QMessageBox::information(this, "isNULL", QString("Папка не была выбрана"));
        return;
    } else {
        filePath=tempDirName;
        read_from_template();
    }

}

void MainWindow::on_pushButton_3_pressed()
{
    ui->lineEdit_2->setEchoMode(QLineEdit::Normal);
}

void MainWindow::on_pushButton_3_released()
{
    ui->lineEdit_2->setEchoMode(QLineEdit::Password);
}

void MainWindow::on_pushButton_2_clicked()
{
    ui->lineEdit_2->setReadOnly(false);
    ui->pushButton_2->setDisabled(true);
    ui->pushButton_3->setEnabled(true);
}

void MainWindow::on_pushButton_4_clicked()
{
    save_to_template();
    //TODO: Implement protection
}


void MainWindow::on_pushButton_5_clicked()
{

}
//////////////////////////////////////////////////////////////////

void MainWindow::read_from_template(){
    QFile file(filePath+"/"+"template.tbl");
    //QMessageBox::information(this, "Info", file.fileName());
    if(!file.exists()){
        QMessageBox::StandardButton reply;
        reply = QMessageBox::question(this, "File does not exist", "Файл template.tbl не существует в рабочей директории. Создать его? При нажатии на 'нет' программа закроется.",
                                      QMessageBox::Yes|QMessageBox::No);
        if (reply == QMessageBox::Yes) {
            file.open(QIODevice::WriteOnly);
            file.close();
        } else {
            qDebug() << "No was clicked";
            this->close();
        }
    } else {
        if(!file.open(QIODevice::ReadOnly)){
            QMessageBox::critical(this, "Error", "Не удалось открыть файл! Программа будет закрыта");
            this->close();
        }
        ui->textEdit->clear();
        QTextStream in(&file);
        QString line = in.readLine();
        if(!line.isEmpty()){
            ui->textEdit->setReadOnly(true);
            ui->lineEdit_2->setReadOnly(true);
            qint8 br_protection = 3;
            QByteArray template_hash = QByteArray::fromHex(line.toUtf8());
            while((br_protection--) > 0) {
                QString input_password = QInputDialog::getText(this, "Info", "В первой строке шаблона содержится хеш пароля. Чтобы изменить шаблон введите пароль");
                QByteArray input_hash = QCryptographicHash::hash(input_password.toUtf8(),QCryptographicHash::Sha256);
                if (template_hash == input_hash){
                    ui->textEdit->setReadOnly(false);
                    ui->lineEdit_2->setText(input_password);
                    ui->pushButton_2->setEnabled(true);
                    ui->pushButton_3->setDisabled(true);
                    break;
                }
            }
            if(br_protection <= 0) QMessageBox::information(this, "Too many attempts", "Вы ошиблись с вводом пароля 3 раза. Вы не сможете изменить этот файл");
        }
        while(!in.atEnd()){
            line = in.readLine();
            if(line.startsWith('#')) continue;
            ui->textEdit->append(line);
        }
        file.close();
    }
}

void MainWindow::save_to_template(){
    QFile file(filePath+"/"+"template.tbl");
    if(!file.exists()){
        QMessageBox::StandardButton reply;
        reply = QMessageBox::question(this, "File does not exist", "Файл template.tbl не существует в рабочей директории. Создать его? При нажатии на 'нет' программа закроется.",
                                      QMessageBox::Yes|QMessageBox::No);
        if (reply == QMessageBox::Yes) {
            file.open(QIODevice::WriteOnly);
            file.close();
        } else {
            qDebug() << "No was clicked";
            this->close();
        }
    }
    //Открываем файл
    if(!file.open(QIODevice::WriteOnly | QIODevice::Text)){
        QMessageBox::critical(this, "Error", "Не удалось открыть файл! Программа будет закрыта");
        this->close();
    }
    QTextStream out(&file);
    //Сохраняем пароль в виде хеша SHA256 в первой строке
    if(!(ui->lineEdit_2->isReadOnly()) && !(ui->lineEdit_2->text().isEmpty())){
        QByteArray hash = QCryptographicHash::hash(ui->lineEdit_2->text().toUtf8(), QCryptographicHash::Sha256);
        out << hash.toHex() << Qt::endl;
    } else out << Qt::endl;
    out << "#First line is autogenerated. Modifying it might make this file uneditable by program" << Qt::endl;
    //Выводим текущие имена и шаблоны в файл
    out << ui->textEdit->toPlainText();
    file.flush();
    file.close();
    //Выставляем соответствующие ограничения на доступ к файлу
    /*
    PACL pOldDACL=NULL;
    PSECURITY_DESCRIPTOR pSD = NULL;
    if(GetNamedSecurityInfo(LPCWSTR(file.fileName().toStdString().c_str()),SE_FILE_OBJECT,DACL_SECURITY_INFORMATION,NULL,NULL, &pOldDACL, NULL, &pSD) != ERROR_SUCCESS){
        QMessageBox::critical(this, "Error!", QString("GetNamedSecurityInfo() for template.tbl failed! Error code:") + QString::number(GetLastError()));
        this->close();
    }*/
    PACL pDACL=NULL;
    EXPLICIT_ACCESS ea;
    ZeroMemory(&ea, sizeof(EXPLICIT_ACCESS));
    ea.grfAccessPermissions = GENERIC_ALL;
    ea.grfAccessMode = SET_ACCESS;
    ea.grfInheritance= NO_INHERITANCE;
    ea.Trustee.TrusteeForm=TRUSTEE_IS_NAME;
    ea.Trustee.ptstrName= L"CREATOR OWNER";
    if(SetEntriesInAcl(1, &ea, NULL, &pDACL) != ERROR_SUCCESS){
        QMessageBox::critical(this, "Error!", QString("SetEntriesInAcl() for template.tbl failed! Error code:") + QString::number(GetLastError()));
        this->close();
    }
    if(SetNamedSecurityInfoA(LPSTR(file.fileName().toStdString().c_str()),SE_FILE_OBJECT, DACL_SECURITY_INFORMATION|PROTECTED_DACL_SECURITY_INFORMATION, NULL, NULL, pDACL, NULL)!=ERROR_SUCCESS){
        QMessageBox::critical(this, "Error", "Ошибка в SetNamedSecurityInfo");
        this->close();
    }
    LocalFree(pDACL);
}




